### PR TITLE
Type specific equal constraints

### LINKED ISSUE #2492: EqualConstraint.Using(StringComparer.*) causes overload ambiguity error
```cs
Assert.That("a", Is.EqualTo("a").Using(StringComparer.Ordinal));
```

fails to compile:

> CS0121 The call is ambiguous between the following methods or properties: `EqualConstraint.Using(IComparer)` and `EqualConstraint.Using<T>(IComparer<T>)`

In fact, `StringComparer` matches four of the overloads. 😆 

We could:

 1. Make the common string comparer selection case easier by adding a new overload to receive a `StringComparison` enum.

 2. Add a new `Using(StringComparer)` overload to disambiguate which forwards to any of the four other `Using` overloads.

 3. Add `Ordinal` and `OrdinalIgnoreCase` modifier properties to `EqualsConstraint` like we have with `IgnoreCase`.

 4. Leave things as they are and make this the minimum code you have to type:
    
    ```cs
    Assert.That("a", Is.EqualTo("a").Using((IComparer<string>)StringComparer.Ordinal));
    ```
    
    Or this, on the rare day that you have `using System.Collections;` already added:
    ```cs
    Assert.That("a", Is.EqualTo("a").Using((IComparer)StringComparer.Ordinal));
    ```

Opinions?

### LINKED ISSUE #4874: IgnoreCase, Seconds and Minutes can be used on numerical constraints
Currently the following is allowed by the compiler, thought it is warned by the NUnit.Analyzer:

```cs
   Assert.That(42,Is.EqualTo(42).IgnoreCase  // Raises NUnit2008 warning.   IgnoreWhitespace
   Assert.That(42,Is.EqualTo(42).IgnoreWhiteSpace);
   Assert.That(42,Is.EqualTo(42).Minutes  // Seconds also allowed.  Doesn't raise any Analyzer warnings.
```

The compiler should report these as errors.

### LINKED ISSUE #4875: String constraints should not allow using Within
Currently NUnit allows (as in compiles) statements like

```cs
Assert.That(somestring,Is.EqualTo("Don't panic").Within(42))  // reports NUnit2047 warning
```

The compiler should report this as an error.

### LINKED ISSUE #4877: Fix EqualTo modifiers for DateTime
The following below is currently allowed, we need to change this to be compiler errors.

So we allow: Assert.That(date1, Is.EqualTo(date2).Within(3).Seconds);
but not: Assert.That(date1, Is.EqualTo(date2).Seconds)
nor: Assert.That(date1, Is.EqualTo(date2).Within(timeSpan).Seconds)

Because the class with tolerance is different than the one without this also doesn't allow repeats like:
Assert.That(date1, Is.EqualTo(date2).Within(1).Hours.Seconds.Within(3))

### PR DESCRIPTION
**Pull request by manfred-brands**
Fixes #4874 
Fixes #4875 
Fixes #4877
Fixes #2492 

This is a merge of #4847, #4848, #4853 with one additional commit restricting combining `Using` with other modifiers.

Once `Using` is selected, a user supplied comparer will be called which does not have access to any `Within` or `IgnoreCase` modifiers.

The classes are now such that once a constraint specific modifiers is selected, the return type will not have the `Using` modifiers.
The reverse also holds, once the `Using` modifiers is used, the return type has no other modifiers.

### CODE REVIEW

**Comment by stevenaw** [comment]
Thanks @manfred-brands 
I won't be able to review tonight, but your notes about `Using` prompted my memory back to https://github.com/nunit/nunit/issues/2492.

I'm not suggesting we account for that here already, but it did occur to me that the type-specialized equalconstraints added here could also make future features of that nature easier too.

**Comment by manfred-brands** [comment]
> your notes about `Using` prompted my memory back to #2492.

It sometimes seems you know all outstanding issues, or have a good search method.

> I'm not suggesting we account for that here already, but it did occur to me that the type-specialized equalconstraints added here could also make future features of that nature easier too.

I added a commit to fix #2492

I also worked on adding an overload to specify StringComparison but that became to big and not sure if it adds anything.

**Comment by stevenaw** [comment]
Thanks @manfred-brands !
It was not my intention to add to the scope here by mentioning that ticket, but your adding it is a great example of how these can improve usability along with type-safety. 

I hope to have time to review within a few days

File: src/NUnitFramework/framework/Constraints/ConstraintExpression.cs [hunk 1]
```diff
@@ -427,6 +427,49 @@ public EqualConstraint EqualTo(object? expected)
...
+            return Append(new EqualTimeBaseConstraint<TimeSpan>(expected, x => x.Ticks));
+        }
+
+        /// <summary>
+        /// Returns a constraint that tests two numbers for equality
+        /// </summary>
+#pragma warning disable CS3024 // Constraint type is not CLS-compliant
+        public EqualNumericConstraint<T> EqualTo<T>(T expected)
+            where T : unmanaged, IConvertible, IEquatable<T>
+        {
```
Comment (OsirisTerje) [inline src/NUnitFramework/framework/Constraints/ConstraintExpression.cs:468]: is this generic constraint `where T : unmanaged` meant to restrict to value types or is it really also for other unmanaged types ?  If the first, any reason not to use struct ?

**Review by OsirisTerje — APPROVED** [review]
Awesome work!  But it is quite a lot.  I'll pull it down and try it out on some projects, that is probably the easiest way to check it out.

**Comment by OsirisTerje** [comment]
Checked #4874 , #4875, and #4877 :  All works :-)

**Comment by OsirisTerje** [comment]
Anything else you want to do here, @manfred-brands ?  Or can we merge?

**Comment by manfred-brands** [comment]
> Anything else you want to do here, @manfred-brands ? Or can we merge?

Are you happy to merge to `main` or do we create a `vNext`?

**Review by OsirisTerje — APPROVED** [review]
lgtm

@manfred-brands Can we merge this in now?

**Comment by manfred-brands** [comment]
@OsirisTerje I have no access to a PC this week to resolve any possible conflicts. Feel free to merge otherwise I'll do it next week.

**Pull request merged**